### ISSUE TITLE
speed and parallelization?

### ISSUE THREAD

**Issue by ignacio82** [vdorie/dbarts#11]
The following code has been running for around 24 hours:

```r
library(dbarts)
library(C50)
data(churn)

test <- bart2(formula = as.numeric(churn) - 1 ~ ., data = churnTrain, verbose = TRUE)
```
I have a few questions:

1. Is this to be expected or am I doing something wrong?
2. How many cores inside a single computer can I use to speed things up. For example, if I were to run this on aws with a big computer with 90 cores, will `dbart`  take advantage of the 90 cores if I set `n.threads=90`. 
3. Instead of using a single computer with many cores, is it possible to use a cluster of computers to run this on parallel?

Thanks again for all the help!

**Comment by vdorie** [comment]
Very strange. On my laptop it takes about 30 seconds. Are you using a version installed from github or on CRAN? I can't remember all of the changes, but I'm often slow to submit to CRAN due to their process.

**Comment by vdorie** [comment]
It's probably a race condition of some kind. Try it with 1 thread, then with 4. If you have more threads than chains they can be used, but aren't generally of benefit unless the number of observations is very large.

I have access to a Windows machine, I'll see if I can't replicate what's causing it to hang.

**Comment by ignacio82** [comment]
This works:
```r
library(dbarts)
library(C50)
library(tictoc)
data(churn)
tic("with one thread")
test <- bart2(formula = as.numeric(churn) - 1 ~ ., data = churnTrain, verbose = TRUE, 
              n.threads = 1L)
toc()
with one thread: 13.162 sec elapsed
```

This does not work:

```r
library(dbarts)
library(C50)
library(tictoc)
data(churn)
tic("with one thread")
test <- bart2(formula = as.numeric(churn) - 1 ~ ., data = churnTrain, verbose = TRUE, 
              n.threads = 4L)
toc()
```

```
> sessionInfo()
R version 3.5.1 (2018-07-02)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 16.04.4 LTS

Matrix products: default
BLAS: /usr/lib/libblas/libblas.so.3.6.0
LAPACK: /usr/lib/lapack/liblapack.so.3.6.0

locale:
[1] C

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
[1] compiler_3.5.1 tools_3.5.1    yaml_2.1.19  
```

**Comment by vdorie** [comment]
Great, thanks! I'll take a look when I can.

**Comment by vdorie** [comment]
Took me a while, but I finally found the culprit. Fixed in 1253d9abd0d641451e8e5f2a7f548afd0ec616c4.
